geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/KryoLazyDensityCoprocessor.scala

```diff
@@ -0,0 +1,168 @@
+import org.locationtech.geomesa.hbase.proto.KryoLazyDensityProto._
+import org.locationtech.geomesa.index.utils.KryoLazyDensityUtils
+import org.locationtech.geomesa.utils.geotools.{GeometryUtils, SimpleFeatureTypes}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.Filter
+import org.slf4j.LoggerFactory
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+class KryoLazyDensityCoprocessor extends KryoLazyDensityService with Coprocessor with CoprocessorService with KryoLazyDensityUtils {
+
+  import KryoLazyDensityCoprocessor._
+
+  private var env : RegionCoprocessorEnvironment = null
+  var sft: SimpleFeatureType = null
+
+  def init(options: Map[String, String], sft: SimpleFeatureType): DensityResult = {
+    this.sft = sft
+    initialize(options, sft)
+  }
+
+    def aggregateResult(sf: SimpleFeature, result: DensityResult): Unit = writeGeom(sf, result)
+
+  @throws[IOException]
+  def start(env: CoprocessorEnvironment) {
+    if (env.isInstanceOf[RegionCoprocessorEnvironment])
+      this.env = env.asInstanceOf[RegionCoprocessorEnvironment]
+    else
+      throw new CoprocessorException("Must be loaded on a table region!")
+  }
+
+  @throws[IOException]
+  def stop(coprocessorEnvironment: CoprocessorEnvironment) {
+  }
+
+  def getService: Service = this
+
+  def getDensity(controller: RpcController, request: KryoLazyDensityProto.DensityRequest, done: RpcCallback[KryoLazyDensityProto.DensityResponse]) {
+    val byteFilterArray = request.getByteFilter.toByteArray
+    var response : DensityResponse = null
+    var scanner : InternalScanner = null
+    try {
+      val scan = new Scan
+      val options: Map[String, String] = deserializeOptions(byteFilterArray)
+      val sft = SimpleFeatureTypes.createType("input", options(SFT_OPT))
+      val serializer = new KryoFeatureSerializer(sft, SerializationOptions.withoutId)
+      val densityResult: DensityResult = this.init(options, sft)
+      scanner = env.getRegion.getScanner(scan)
+      val results = new java.util.ArrayList[Cell]
+      var hasMore = false
+      do {
+        hasMore = scanner.next(results)
+        for (cell <- results) {
+          val sf = serializer.deserialize(cell.getValueArray, cell.getValueOffset, cell.getValueLength)
+          aggregateResult(sf, densityResult)
+        }
+        results.clear()
+      } while (hasMore)
+
+      val result: Array[Byte] = KryoLazyDensityUtils.encodeResult(densityResult)
+      response = DensityResponse.newBuilder.setSf(ByteString.copyFrom(result)).build
+    } catch {
+      case ioe: IOException => {
+        ResponseConverter.setControllerException(controller, ioe)
+      }
+      case cnfe: ClassNotFoundException => {
+        logger.error(Throwables.getStackTraceAsString(cnfe))
+      }
+      case dse: DeserializationException => {
+        logger.error(Throwables.getStackTraceAsString(dse))
+      }
+    } finally IOUtils.closeQuietly(scanner)
```

Comment:
i think we have a new way to close things in utils instead of using this: See ```geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/io/SafeClose.scala``` maybe use close with logging